lib/fastimage.rb

```diff
@@ -568,6 +583,153 @@ def parse_size_for_ico
   end
   alias_method :parse_size_for_cur, :parse_size_for_ico
 
+  class Heic # :nodoc:
+    def initialize(stream)
+      @stream = stream
+    end
+
+    def width_and_height
+      @max_size = nil
+      @primary_box = nil
+      @ipma_boxes = []
+      @ispe_boxes = []
+      @final_size = nil
+
+      catch :finish do
+        read_boxes!
+      end
+
+      @final_size
+    end
+
+    private
+
+    def read_boxes!(max_read_bytes = nil)
+      end_pos = max_read_bytes.nil? ? nil : @stream.pos + max_read_bytes
+      index = 0
+
+      loop do
+        end_pos.nil? || @stream.pos < end_pos or
+          return
```

Comment:
One thing: I find this a bit hard to read.
It's mostly a matter of taste, but I find the postfix use of 'or' harder to read than an 'if'

    return if end_pos && @stream.pos >= end_pos

Likewise, I prefer this to using 'or' even in the simple case: this is clearer to me

    throw :finish if box_size < 12